odl/solvers/iterative/iterative.py

```diff
@@ -290,6 +289,106 @@ def conjugate_gradient_normal(op, x, rhs, niter=1, callback=None):
             callback(x)
 
 
+def conjugate_gradient_nonlinear(f, x, rhs, niter=1, nreset=0,
+                                 line_search=None, tol=1e-16, beta_method='FR',
+                                 callback=None):
+    """Conjugate gradient for nonlinear problems.
+
+    The method is described in a
+    `Wikipedia article
+    <https://en.wikipedia.org/wiki/Nonlinear_conjugate_gradient_method>`_.
+
+    Parameters
+    ----------
+    op : `Functional`
+        Operator in the inverse problem. If not linear, it must have
+        an implementation of `Operator.derivative`, which
+        in turn must implement `Operator.adjoint`, i.e.
+        the call ``op.derivative(x).adjoint`` must be valid.
+    x : ``op.domain`` element
+        Vector to which the result is written. Its initial value is
+        used as starting point of the iteration, and its values are
+        updated in each iteration step.
+    rhs : ``op.range`` element
+        Right-hand side of the equation defining the inverse problem
+    niter : int
+        Number of iterations per reset.
+    nreset : int, optional
+        Number of times the solver should be reset. Default: no reset.
+    line_search : float or `LineSearch`, optional
+        Strategy to choose the step length. If a float is given, uses it as a
+        fixed step length. Default: `BacktrackingLineSearch`
+    tol : float, optional
+        Tolerance that should be used for terminating the iteration.
+    beta_method : {'FR', 'PR', 'HS', 'DY'}
+        Method to calculate ``beta`` in the iterates. TODO
```

Comment:
Yeah I'll add that but that goes up in notes, not here since that would clog this.